### ISSUE TITLE
Introduce semantic versioning compliance specification for Ballerina APIs

### ISSUE THREAD

**Issue by NipunaRanasinghe** [ballerina-platform/ballerina-spec#1070]
**Description:**

Currently the Ballerina library(package) developers can change the package version at their will, because currently we don’t have any builtin support for a validation/enforcing mechanism to ensure semantic versioning compliance between consecutive package releases. And This might end up having packages in our central that are compatible only by its version, but not by the implementation. 

Therefore we are having this idea of introducing 
 - a semver compliance specification for Ballerina packages, which precisely determines which kind of changes in the source code can affect the package version
 - a tool which can validate the package changes against its version, conforming to the aforementioned specification. 

Therefore currently I have started writing the aforementioned specification which is attached below.  @jclark Any of your thoughts and feedback on the proposed specification is highly appreciated.

Proposed Specification [Draft]  -https://docs.google.com/document/d/1YowcycsEcO0K67sYBAb_Gz1slFVf5CnCkgkOnMMKcSo/edit?usp=sharing

**Suggested Labels:**

   
**Code sample that shows issue:**

**Related Issues:**
https://github.com/ballerina-platform/ballerina-lang/issues/32373
https://github.com/ballerina-platform/ballerina-lang/issues/32368

**Comment by jclark** [comment]
That draft is not public: please don't put links to non-public documents in public issues. Please turn it into Markdown.

Which changes cause incompatibilities will depend in some respects on how the compiler implements things.

**Comment by NipunaRanasinghe** [comment]
> That draft is not public: please don't put links to non-public documents in public issues. Please turn it into Markdown.

Sorry for any inconvenience caused @jclark. Please find the link below for the public document. I'll share the markdown as soon as this is proposal is reviewed and finalised. 

Proposed Specification [Draft] -https://docs.google.com/document/d/1STs91QTXrvivfcowGltOtoYCRlgJheGPay56G3cWuxM/edit?usp=sharing

**Comment by NipunaRanasinghe** [comment]
@jclark @sameerajayasoma We have started the initial implementation of the tool as a new standalone command and, completed the first milestone(support for functional APIs) with https://github.com/ballerina-platform/ballerina-lang/pull/36259 (including a design review and code reviews).  Currently the tool can 
1. compare the local changes with any published package version available in central
2. suggest the next package version based on the source code changes and, 
3. list-down/export the list of source code changes between the local code and any published package version

## Sample 
Please refer to the sample below, which demonstrate the functionalities of the new command on the `ballerina/io` package.

```
i. version suggestions

$bal semver             
checking for the latest compatible release version available in central...

current version: 1.2.1
compatibility impact (compared with the release version '1.2.1'): backward-incompatible changes detected
suggested version: 2.0.0
ii. version suggestions with the list of detected source code changes

$bal semver --show-diff
checking for the latest compatible release version available in central...

================================================================
 Comparing version '1.2.1'(local) with version '1.2.1'(central) 
================================================================
[+-] package 'io' is modified [version impact: MAJOR]
  [+-] module 'io' is modified [version impact: MAJOR]
    [++] function 'printlnNew' is added [version impact: MINOR]
    [+-] function 'externPrintln' is modified [version impact: PATCH]
      [++] new rest parameter 'vlues' is added [version impact: MINOR]
      [++] rest parameter 'values' is removed [version impact: MAJOR]
    [+-] function 'externPrint' is modified [version impact: PATCH]
      [++] new rest parameter 'vlues' is added [version impact: MINOR]
      [++] rest parameter 'values' is removed [version impact: MAJOR]
    [+-] function 'println' is modified [version impact: MAJOR]
      [+-] documentation is modified [version impact: PATCH]
      [--] 'isolated' qualifier is removed [version impact: AMBIGUOUS]
      [++] 'transactional' qualifier is added [version impact: AMBIGUOUS]
      [++] new required parameter 'a' is added [version impact: MAJOR]
      [++] new defaultable parameter 'b' is added [version impact: MINOR]
      [+-] parameter type changed from 'Printable' to 'string' [version impact: AMBIGUOUS]
      [++] return type is added [version impact: MAJOR]
      [+-] function body is modified [version impact: PATCH]

current version: 1.2.1
compatibility impact (compared with the release version '1.2.1'): backward-incompatible changes detected
suggested version: 2.0.0
```
## Command Options
Please find all the command options supported by the tool

```
 -d, --show-diff
     Display the detailed information of source code differences between
     the compared package versions

 -c <version>, --compare-version <version>
     Analyze SemVer compliance of the local changes against the specified
     release version.
```

Therefore we'd be really grateful if you let us know your thoughts and, if you have any different suggestions on the tool name, command name, command options, etc.

**Comment by sameerajayasoma** [comment]
+1 for the command name `bal semver`.

Here is my understanding of what this tool does: 
- Compares the current package with the latest compatible released version in Ballerina central.
- Output an error message if the current package version is inconsistent with the changes. I.e., Say there are breaking changes in the current package, but the version is just a minor bump, then this tool recognizes that as an error.  
- If the current package version is consistent with the changes, it doesn't output an error. 
- The command output can be used by CI tools to check semver compliance of the new version and fails builds if required. 
- Optionally, this command prints a detailed list of changes 

@NipunaRanasinghe, please confirm whether my understanding aligns with what you are working on :) 

I will comment on command options and the output afterward.

**Comment by NipunaRanasinghe** [comment]
@sameerajayasoma Yes you are correct! :) And in addition to what you've already mentioned, the tool is expected to have the following functionalities as well

- If the current package version is inconsistent with the changes, the tool will suggest a **minimum compatible version** for the package based on the impact of the changes.
- The tool will also detect and inform the user if package version is loosely defined (i.e. there are  only patch compatible changes in the current package, but the major version is bumped).
- The tool can serialize the list (the tree, to be more precise) of changes and export to a file (i.e. json) so that users/CI tools can analyse the version impact of individual changes if required. (core functionality is already implemented, just need to introduce a new flag)
